File: ui-ngx/src/app/core/http/audit-log.service.ts
// SPDX-FileCopyrightText: Copyright The Thingsboard Authors
// SPDX-License-Identifier: Apache-2.0
import { Injectable } from '@angular/core';
import { createDefaultHttpOptions, RequestConfig } from './http-utils';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { TimePageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';
import { AuditLog, AuditLogFilter } from '@shared/models/audit-log.models';
import { EntityId } from '@shared/models/id/entity-id';

@Injectable({
  providedIn: 'root'
})
export class AuditLogService {

  constructor(
    private http: HttpClient
  ) { }

  public getAuditLogs(pageLink: TimePageLink, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogs(pageLink: TimePageLink, filters: AuditLogFilter, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogs(
    pageLink: TimePageLink,
    filtersOrConfig?: AuditLogFilter | RequestConfig,
    config?: RequestConfig
  ): Observable<PageData<AuditLog>> {
    return this.http.get<PageData<AuditLog>>(
      `/api/audit/logs${pageLink.toQuery()}`,
      createDefaultHttpOptions(filtersOrConfig, config)
    );
  }

  public getAuditLogsByCustomerId(customerId: string, pageLink: TimePageLink, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByCustomerId(customerId: string, pageLink: TimePageLink, filters: AuditLogFilter, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByCustomerId(
    customerId: string,
    pageLink: TimePageLink,
    filtersOrConfig?: AuditLogFilter | RequestConfig,
    config?: RequestConfig
  ): Observable<PageData<AuditLog>> {
    return this.http.get<PageData<AuditLog>>(
      `/api/audit/logs/customer/${customerId}${pageLink.toQuery()}`,
      createDefaultHttpOptions(filtersOrConfig, config)
    );
  }

  public getAuditLogsByUserId(userId: string, pageLink: TimePageLink, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByUserId(userId: string, pageLink: TimePageLink, filters: AuditLogFilter, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByUserId(
    userId: string,
    pageLink: TimePageLink,
    filtersOrConfig?: AuditLogFilter | RequestConfig,
    config?: RequestConfig
  ): Observable<PageData<AuditLog>> {
    return this.http.get<PageData<AuditLog>>(
      `/api/audit/logs/user/${userId}${pageLink.toQuery()}`,
      createDefaultHttpOptions(filtersOrConfig, config)
    );
  }

  public getAuditLogsByEntityId(entityId: EntityId, pageLink: TimePageLink, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByEntityId(entityId: EntityId, pageLink: TimePageLink, filters: AuditLogFilter, config?: RequestConfig): Observable<PageData<AuditLog>>;
  public getAuditLogsByEntityId(
    entityId: EntityId,
    pageLink: TimePageLink,
    filtersOrConfig?: AuditLogFilter | RequestConfig,
    config?: RequestConfig
  ): Observable<PageData<AuditLog>> {
    return this.http.get<PageData<AuditLog>>(
      `/api/audit/logs/entity/${entityId.entityType}/${entityId.id}${pageLink.toQuery()}`,
      createDefaultHttpOptions(filtersOrConfig, config)
    );
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/audit/AuditLogService.java
// SPDX-FileCopyrightText: Copyright The Thingsboard Authors
// SPDX-License-Identifier: Apache-2.0
package org.thingsboard.server.dao.audit;

import com.google.common.util.concurrent.ListenableFuture;
import org.thingsboard.server.common.data.HasName;
import org.thingsboard.server.common.data.audit.ActionType;
import org.thingsboard.server.common.data.audit.AuditLog;
import org.thingsboard.server.common.data.id.CustomerId;
import org.thingsboard.server.common.data.id.EntityId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.id.UserId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.TimePageLink;

import java.util.List;

public interface AuditLogService {

    PageData<AuditLog> findAuditLogsByTenantIdAndCustomerId(TenantId tenantId, CustomerId customerId, List<ActionType> actionTypes, TimePageLink pageLink);

    PageData<AuditLog> findAuditLogsByTenantIdAndUserId(TenantId tenantId, UserId userId, List<ActionType> actionTypes, TimePageLink pageLink);

    PageData<AuditLog> findAuditLogsByTenantIdAndEntityId(TenantId tenantId, EntityId entityId, List<ActionType> actionTypes, TimePageLink pageLink);

    PageData<AuditLog> findAuditLogsByTenantId(TenantId tenantId, List<ActionType> actionTypes, TimePageLink pageLink);

    <E extends HasName, I extends EntityId> ListenableFuture<Void> logEntityAction(
            TenantId tenantId,
            CustomerId customerId,
            UserId userId,
            String userName,
            I entityId,
            E entity,
            ActionType actionType,
            Exception e, Object... additionalInfo);
}
